Add ExCL H100 environment

sethrj opened the pull request:
This adds the environment and build files for running on the `hudson` ExCL machine with an H100. All tests build and pass with the `reldeb` and `reldeb-orange` profiles.

sethrj committed:
Add hudson environment

sethrj committed:
WIP: try to get cuda to compile

sethrj committed:
WTF thrust

sethrj committed:
Revert things

sethrj committed:
Add ccache, disable device debug

sethrj committed:
Fix missing libcudart link errors

sethrj committed:
Add LC_ALL export

LSchwiebert commented on scripts/cmake-presets/hudson.json:
```diff
@@ -0,0 +1,94 @@
+        "CELERITAS_DEVICE_DEBUG":{"type": "BOOL", "value": "OFF"},
+        "CELERITAS_USE_VecGeom": {"type": "BOOL", "value": "OFF"}
+      }
+    }
```
If it makes sense to everyone else, leave it, but I'm having trouble distinguishing between the displayName strings for reldeb-orange and debug-orange. Perhaps change the displayName for reldeb-orange to:

displayName": "Build with basic optimizations and debug assertions and ORANGE geometry"

I somehow quoted the wrong lines. Look a few lines above that.

sethrj replied:
I copied these from JLSE and it could definitely use improvement, I'll fix.

pcanal commented:
Is this confirmation that the change 'already' made to `CudaRdcUtils` is solving the linking problem?

sethrj commented:
@pcanal What change? On develop the build fails; with the proposed change the build passes.

pcanal commented:
I was asking whether [`b55b999` (#2107)](https://github.com/celeritas-project/celeritas/pull/2107/commits/b55b999a3997b9e63a91cbe4853f532211ad4bdf) was the only change in CudaRdcUtils that was needed or if there was more investigation needed.

sethrj commented:
Aha, that was indeed the only change needed. Did you want any more details, or are you OK to create a new RDC version? Did I see there was an update to the RdcUtils in vecgeom, or is that not needed?

sethrj commented:
I'll wait for #2110 since it has a necessary fix that I independently figured out for this one...

sethrj committed:
Write clangd for hudson vscode usage

pcanal commented:
>  Did you want any more details,

No. we are good to go.

sethrj committed:
Add informative message when building on login node

sethrj committed:
Test spack directory for readability

pre-commit-ci[bot] committed:
Apply fixes from pre-commit hooks: see detailed commit message →

This commit means that you must (please!) install pre-commit
on your development machine and run `pre-commit install --install-hooks`.
For more information, see
https://celeritas-project.github.io/celeritas/user/development/style.html#formatting

Autogenerated: https://pre-commit.ci

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into hudson-env

sethrj committed:
Improve cmake preset descriptions

sethrj commented:
ok @elliottbiondo @LSchwiebert @Rashika-Gupta this is ready do go, at least one or two of you should have access to the system now 🙃

sethrj commented:
@pcanal I mistakenly thought that, since the building and linking passed, your V13 updates fixed it. However, sadly, what used to be a link failure is now a runtime failure :( Is it feasible to require the exe linking to require all symbols be resolved? This seems like a major regression.

To fix the build, I still have to add an extra cuda toolkit link to the middle library to compile:
```diff
diff --git a/cmake/external/CudaRdcUtils.cmake b/cmake/external/CudaRdcUtils.cmake
index b016abb4..060b0230 100644
--- a/cmake/external/CudaRdcUtils.cmake
+++ b/cmake/external/CudaRdcUtils.cmake
@@ -527,6 +527,8 @@ function(cuda_rdc_add_library target)
     CUDA_RESOLVE_DEVICE_SYMBOLS OFF # We really don't want nvlink called.
     EXPORT_PROPERTIES "CUDA_RUNTIME_LIBRARY;CUDA_RDC_LIBRARY_TYPE;CUDA_RDC_FINAL_LIBRARY;CUDA_RDC_MIDDLE_LIBRARY;CUDA_RDC>
   )
+  # Ensure middle library and its dependents see the `-lcudart` library path
+  target_link_libraries(${target} PUBLIC CUDA::toolkit)

   if(CUDA_RDC_LINKER_SUPPORTS_ALLOW_SHLIB_UNDEFINED)
     target_link_options(${target} PRIVATE LINKER:--allow-shlib-undefined)
```

The difference the extra link makes is shown below using [libtree](https://github.com/haampie/libtree).

```diff
$ diff -u vg-libtree-hudsonenv.txt vg-libtree-hudsonenv-new.txt
--- vg-libtree-hudsonenv.txt	2025-11-19 07:17:20.238351226 -0500
+++ vg-libtree-hudsonenv-new.txt	2025-11-19 07:20:10.433398212 -0500
@@ -14,8 +14,8 @@
 │   │   ├── libgtest.so.1.17.0 [runpath]
 │   │   │   ├── libm.so.6 not found: RPATH (see below) LD_LIBRARY_PATH (empty) RUNPATH (empty) /etc/ld.so.conf (empty)
 │   │   │   └── libc.so.6 not found: RPATH (see below) LD_LIBRARY_PATH (empty) RUNPATH (empty) /etc/ld.so.conf (empty)
-│   │   ├── libm.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
-│   │   └── libc.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
+│   │   ├── libm.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
+│   │   └── libc.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
 │   ├── libgeocel.so [runpath]
 │   │   ├── libg4vg.so [runpath]
 │   │   │   ├── libG4gdml.so [rpath]
@@ -169,9 +169,9 @@
 │   │   ├── libG4global.so (collapsed) [runpath]
 │   │   ├── libCLHEP-2.4.7.1.so (collapsed) [runpath]
 │   │   ├── libxerces-c-3.3.so (collapsed) [runpath]
-│   │   ├── libcudart.so.12 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
-│   │   ├── libm.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
-│   │   └── libc.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
+│   │   ├── libcudart.so.12 (collapsed) [runpath]
+│   │   ├── libm.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
+│   │   └── libc.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
 │   ├── libgtest.so.1.17.0 (collapsed) [runpath]
 │   ├── libcorecel.so (collapsed) [runpath]
 │   ├── libG4geometry.so (collapsed) [runpath]
@@ -183,14 +183,14 @@
 │   ├── libtestcel_harness.so (collapsed) [runpath]
 │   ├── libcorecel.so (collapsed) [runpath]
 │   ├── libgtest.so.1.17.0 (collapsed) [runpath]
-│   ├── libm.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
-│   └── libc.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
+│   ├── libm.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
+│   └── libc.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
 ├── libtestcel_harness.so (collapsed) [runpath]
 ├── libgeocel.so (collapsed) [runpath]
 ├── libvecgeom.so (collapsed) [runpath]
 ├── libcorecel.so (collapsed) [runpath]
 ├── libgtest.so.1.17.0 (collapsed) [runpath]
-├── libcudart.so.12 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
-├── libm.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
-└── libc.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
+├── libcudart.so.12 (collapsed) [runpath]
+├── libm.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
+└── libc.so.6 not found: RPATH (empty) LD_LIBRARY_PATH (empty) RUNPATH (see below) /etc/ld.so.conf (empty)
```

```
RUNPATH:"/scratch/s3j/build/celeritas-reldeb/lib":"/auto/projects/celeritas/spack/var/spack/environments/celeritas/.spack-env/view/lib":"/auto/projects/celeritas/spack/var/spack/environments/celeritas/.spack-env/view/lib64"
RPATH:"/auto/projects/celeritas/spack/opt/spack/ubuntu24.04/zen4/googletest/1.17.0/5e7hlac/lib":"/auto/projects/celeritas/spack/opt/spack/ubuntu24.04/zen4/googletest/1.17.0/5e7hlac/lib":"/auto/projects/celeritas/spack/opt/spack/ubuntu24.04/zen4/gcc-runtime/13.3.0/u6ksoly/lib"
```

sethrj committed:
Fix linking

sethrj committed:
Fix git dir and message

sethrj committed:
Add informative message

sethrj committed:
Delete stray comment

sethrj committed:
Update pre-commit

Rashika-Gupta reviewed (approved):
Thank you, Seth :)

sethrj committed:
Don't load all of spack when just setting up build environment

sethrj committed:
Merge branch 'develop' into hudson-env
